### PR TITLE
More detemplating (citystate and notlast)

### PR DESCRIPTION
**Pull request by SansPapyrus683**
As for `notlast`, I'm not sure why there's a case for if there's 7 cows- could someone explain it to me?

All solutions have been tested.

### CODE REVIEW

File: solutions/USACO dec16-citystate.mdx [hunk 1]
```diff
@@ -20,222 +20,26 @@ As the editorial mentions, we can store the number of times a given four-letter
...
+        string state;
+        read >> city >> state;
+		city = city.substr(0, 2);
+        if (city != state) {
+            total += city_state[state + city];
+        }
+		city_state[city + state]++;
 	}
-	ps(ans);
+	ofstream ("citystate.out") << total << endl;
```
Comment (dongliu0426) [inline solutions/USACO dec16-citystate.mdx:42]: spacing o_O

File: solutions/USACO dec16-citystate.mdx [hunk 2]
```diff
@@ -245,228 +49,35 @@ A faster solution is to convert each string into an integer in the range $[0,26^
...
-		if (city != state) ans += cnt[index(state+city)];
-		cnt[index(city+state)] ++;
+		if (city != state) {
+            total += seen[index(state+city)];
+        }
+		seen[index(city + state)]++;
 	}
-	ps(ans);
+	ofstream("citystate.out") << total << endl;
 }
```
Comment (dongliu0426) [inline solutions/USACO dec16-citystate.mdx:81]: O_o

File: solutions/USACO jan17-not-last.mdx [hunk 3]
```diff
@@ -36,230 +36,42 @@ Henrietta 4
...
-		ps(cows[ind].s);
-		exit(0);
+
+	if (ind < cows.size() && (ind + 1 == cows.size() || cows[ind].first != cows[ind + 1].first)) {
+		ofstream("notlast.out") << cows[ind].second << endl;
+		return 0;
 	}
-	ps("Tie");
+	ofstream("notlast.out") << "Tie" << endl;
 }
```
Comment (dongliu0426) [inline solutions/USACO jan17-not-last.mdx:75]: O_o

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

**Comment by dongliu0426** [comment]
please use tabs

**Comment by SansPapyrus683** [comment]
I think the weird spacing is also because of the indentation. I'll fix it right away.

**Review by dongliu0426 — APPROVED** [review]

**Comment by bqi343** [comment]
> I'm not sure why there's a case for if there's 7 cows- could someone explain it to me?

> Note that M=0 if one of the seven cows is completely absent from the milking log, since this cow would have produced no milk.

**Comment by SansPapyrus683** [comment]
Not to nag, but can this be merged yet?

File: solutions/USACO dec16-citystate.mdx [hunk 4]
```diff
@@ -245,228 +50,35 @@ A faster solution is to convert each string into an integer in the range $[0,26^
...
+	long long total = 0;
+	for (int i = 0; i < N; i++) {
+		string city;
+		string state;
+		read >> city >> state;
 		city = city.substr(0,2);
-		if (city != state) ans += cnt[index(state+city)];
-		cnt[index(city+state)] ++;
+		if (city != state) {
+			total += seen[index(state+city)];
```
Comment (maggie-j-liu) [inline solutions/USACO dec16-citystate.mdx:77]:
```suggestion
			total += seen[index(state + city)];
```

File: solutions/USACO jan17-not-last.mdx [hunk 5]
```diff
@@ -36,230 +36,44 @@ Henrietta 4
...
 	}
-	V<pair<int,str>> cows;
-	trav(t,tot) cows.pb({t.s,t.f});
-	sort(all(cows));
+	sort(cows.begin(), cows.end());
+
 	int ind = 0;
-	if (sz(cows) == 7) {
-		while (ind < sz(cows) && cows[ind].f == cows[0].f) ind ++;
+	// only move the pointer if all cows produced some milk- any unmentioned cows will have produced 0 milk
```
Comment (maggie-j-liu) [inline solutions/USACO jan17-not-last.mdx:65]: This line is too long, put a line break after the dash?

File: solutions/USACO jan17-not-last.mdx [hunk 6]
```diff
@@ -36,230 +36,44 @@ Henrietta 4
...
+	if (cows.size() == COW_NUM) {
+		while (ind < cows.size() && cows[ind].first == cows[0].first) {
+			ind++;
+		}
 	}
-	if (ind < sz(cows) && (ind+1 == sz(cows) || cows[ind].f != cows[ind+1].f)) {
-		ps(cows[ind].s);
-		exit(0);
+
+	if (ind < cows.size() && (ind + 1 == cows.size() || cows[ind].first != cows[ind + 1].first)) {
```
Comment (maggie-j-liu) [inline solutions/USACO jan17-not-last.mdx:72]: Also too long, break it into two lines

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/USACO jan17-not-last.mdx (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/USACO jan17-not-last.mdx:72]: I mean, it isn't long enough for the website to wrap it, so I think this one is fine.

Comment (maggie-j-liu) [reply solutions/USACO jan17-not-last.mdx:73]: ?? Maybe you have a larger screen than me, but it wraps on my screen. It's longer than 80 characters, and lines should not exceed 80 characters as mentioned [here](https://usaco.guide/general/contributing?lang=cpp#code-conventions).

Comment (SansPapyrus683) [reply solutions/USACO jan17-not-last.mdx:73]: Ah, I see. I'll break it up then.

**Review by maggie-j-liu — APPROVED** [review]

**Review by frodakcin — APPROVED** [review]
Looks good (actually I just barely skimmed it, and the code looks fine)

I think you can credit yourself in the module authors list instead of in the code?

**Comment by SansPapyrus683** [comment]
I mean I just did some detemplating which isn't that big of a deal, so I'd rather just leave a small watermark of sorts.

**Comment by tachithanhdanh** [comment]
I came here just to put the [flag](https://github.com/cpinitiative/usaco-guide/issues/806)

**Pull request merged**